miasm2/analysis/mem.py

```diff
@@ -0,0 +1,1259 @@
+                         " impossible. Use MemSizedArray instead.")
+
+    def __repr__(self):
+        return "[%r, ...] [%r]" % (self[0], self._field_type)
+
+
+class MemSizedArray(MemArray):
+    """A fixed size MemArray. Its additional arg represents the @array_len (in
+    number of elements) of this array.
+
+    This type is dynamically sized. Use mem_sized_array_type to generate a
+    fixed @field_type and @array_len array which has a static size.
+    """
+    _array_len = None
+
+    def __init__(self, vm, addr=None, field_type=None, array_len=None):
+        # Set the length before anything else to allow get_size() to work for
+        # allocation
+        if self._array_len is None:
+            self._array_len = array_len
+        super(MemSizedArray, self).__init__(vm, addr, field_type)
+        if self._array_len is None or self._field_type is None:
+            raise NotImplementedError(
+                "Provide field_type and array_len to instanciate this class, "
+                "or generate a subclass with mem_sized_array_type.")
+
+    @property
+    def array_len(self):
+        """The length, in number of elements, of this array."""
+        return self._array_len
+
+    def sizeof(cls):
+        raise ValueError("MemSizedArray is not statically sized. Use "
+                         "mem_sized_array_type to generate a type that is.")
+
+    def get_size(self):
+        return self._array_len * self._field_type.size()
+
+    def _normalize_idx(self, idx):
+        if idx < 0:
+            return self.get_size() - idx
+        return idx
+
+    def _check_bounds(self, idx):
+        if not isinstance(idx, int) and not isinstance(idx, long):
+            raise ValueError("index must be an int or a long")
+        if idx < 0 or idx >= self.get_size():
+            raise IndexError("Index %s out of bounds" % idx)
+
+    def __iter__(self):
+        for i in xrange(self._array_len):
+            yield self[i]
+
+    def raw(self):
+        return self._vm.get_mem(self.get_addr(), self.get_size())
+
+    def __repr__(self):
+        item_reprs = [repr(item) for item in self]
+        if self.array_len > 0 and '\n' in item_reprs[0]:
+            items = '\n' + indent(',\n'.join(item_reprs), 2) + '\n'
+        else:
+            items = ', '.join(item_reprs)
+        return "[%s] [%r; %s]" % (items, self._field_type, self._array_len)
+
+    def __eq__(self, other):
+        # Special implementation to handle dynamic subclasses
+        return isinstance(other, MemSizedArray) and \
+                self._field_type == other._field_type and \
+                self._array_len == other._array_len and \
+                str(self) == str(other)
+
+
+def mem_array_type(field_type):
+    """Generate a MemArray subclass that has a fixed @field_type. It allows to
+    instanciate this class with only vm and addr argument, as are standard
+    MemStructs.
+    """
+    cache_key = (field_type, None)
```

Comment:
I may be wrong, but I think the trap with using `.setdefault` here is to mask the original class name with a newly instanciated one (the `default` of `setdefault`) and returning the class that was in the cache. I think it would work, but I do not really like the namespace pollution it may produce (though I may already have name collisions in my generated classes...).

So what do you think, `setdefault` or not?